a_test.go

```diff
@@ -183,42 +185,374 @@ func TestA_uses_Logger_dynamic_interface(t *testing.T) {
+			if parallel {
+				expectedFatalCalls = len(childTasks)
+			}
+
+			assertTrue(t, loggerSpy.called, "logger call")
+			assertEqual(t, loggerSpy.fatalCallCount, expectedFatalCalls, "fatal calls count")
+		})
+	}
+}
+
+func TestA_WithContextSkipped(t *testing.T) {
+	t.Parallel()
+	type ctxKeyT struct{}
+	var ctxKey ctxKeyT
+
+	for _, parallel := range []bool{false, true} {
+		parallel := parallel
+		t.Run("Parallel="+strconv.FormatBool(parallel), func(t *testing.T) {
+			t.Parallel()
+
+			flow := &goyek.Flow{}
+
+			flow.SetOutput(io.Discard)
+			loggerSpy := &helperLoggerSpy{}
+			flow.SetLogger(loggerSpy)
+
+			depsTask := flow.Define(goyek.Task{
+				Name: "deps",
+				Action: func(a *goyek.A) {
+					a.Cleanup(onceCall(t, a.Name()+" cleanup"))
+					onceCall(t, a.Name())()
+				},
+			})
+
+			childTasks := make(goyek.Deps, 10)
+			childTasksResults := make([]bool, len(childTasks))
+
+			task := flow.Define(goyek.Task{
+				Name: "task",
+				Action: func(a *goyek.A) {
+					a.Cleanup(onceCall(t, a.Name()+" cleanup"))
+					onceCall(t, a.Name())()
+
+					k, ok := a.Context().Value(ctxKey).(int)
+					if !ok {
+						return
+					}
+
+					childTasksResults[k] = true
+
+					a.Skip(k)
+					a.Helper()
+				},
+				Deps: goyek.Deps{depsTask},
+			})
+
+			for i := 0; i < len(childTasks); i++ {
+				i := i
+				childTasks[i] = flow.Define(goyek.Task{
+					Name: "child task " + strconv.Itoa(i),
+					Action: func(a *goyek.A) {
+						a.Cleanup(onceCall(t, a.Name()+" cleanup"))
+						onceCall(t, a.Name())()
+
+						prevCtx := a.Context()
+						newA := a.WithContext(context.WithValue(prevCtx, ctxKey, i))
+						assertEqual(t, newA.Name(), a.Name(), "name changed for "+a.Name())
+						assertEqual(t, a.Context(), prevCtx, "context changed for "+a.Name())
+
+						task.Action()(newA)
+					},
+					Deps:     goyek.Deps{depsTask},
+					Parallel: parallel,
+				})
+			}
+
+			main := flow.Define(goyek.Task{
+				Name: "main",
+				Deps: childTasks,
+			})
+
+			err := flow.Execute(context.Background(), []string{main.Name()})
+			assertPass(t, err, "flow execute")
+
+			for index, result := range childTasksResults {
+				assertTrue(t, result, "child result "+strconv.Itoa(index)+" is false")
+			}
+
+			assertTrue(t, loggerSpy.called, "logger call")
+			assertEqual(t, loggerSpy.skipCallCount, len(childTasks), "skip calls count")
+		})
+	}
+}
+
+func TestA_WithContextNilCtx(t *testing.T) {
+	t.Parallel()
+
+	flow := &goyek.Flow{}
+
+	flow.SetOutput(io.Discard)
+	loggerSpy := &helperLoggerSpy{}
+	flow.SetLogger(loggerSpy)
+
+	task := flow.Define(goyek.Task{
+		Name: "test",
+		Action: func(a *goyek.A) {
+			a.WithContext(nil)
+		},
+	})
+
+	assertFail(t, flow.Execute(context.Background(), []string{task.Name()}), "must contain error")
+}
+
+func onceCall(t *testing.T, name string) func() {
+	t.Helper()
+
+	var callsCount int
+	var callsCountM sync.Mutex
+
+	t.Cleanup(func() {
+		assertEqual(t, callsCount, 1, "unexpected number of calls '"+name+"'")
```

Comment:
It is hard to determine if reading `callsCount` is always thread safe here.

You should use `atomic.Int32` or add `callsCountM.Lock()` in the cleanup:

```suggestion
		callsCountM.Lock()
		defer callsCountM.Unlock()
		assertEqual(t, callsCount, 1, "unexpected number of calls '"+name+"'")
```

From https://go.dev/ref/mem

> Don't be clever. 

:wink:

Reply:
PS. There is an edge case where it could race. When the function returned by `onceFunc` would be called in seperate goroutine. Then it can run concurrently with test cleanup thus produce a race condition.